File: Lab7/app/src/main/java/com/example/lab7/MainActivity.java
package com.example.lab7;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.os.Message;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.SeekBar;
import android.widget.Toast;

public class MainActivity extends AppCompatActivity
{
    private int progressHare = 0;
    private int progressTortoise = 0;

    private Button btnStart;
    private SeekBar sbHare, sbTortoise;

    @Override
    protected void onCreate(Bundle savedInstanceState)
    {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        btnStart = findViewById(R.id.btnStart);
        sbHare = findViewById(R.id.sbHare);
        sbTortoise = findViewById(R.id.sbTortoise);

        btnStart.setOnClickListener(v ->
        {
                btnStart.setEnabled(false);
                progressHare = 0;
                progressTortoise = 0;
                sbHare.setProgress(0);
                sbTortoise.setProgress(0);
                runHare();
                runTortoise();

        });
    }

    private final Handler handler = new Handler(Looper.myLooper(), new Handler.Callback()
    {
        @Override
        public boolean handleMessage(@NonNull Message msg)
        {
            Log.e("Hare", String.valueOf(progressHare));
            Log.e("Tortoise", String.valueOf(progressTortoise));

            if (msg.what == 1)
                sbHare.setProgress(progressHare);
            else if (msg.what == 2)
                sbTortoise.setProgress(progressTortoise);

            if(progressHare >= 100 && progressTortoise < 100)
            {
                Toast.makeText(MainActivity.this, "The Hare wins", Toast.LENGTH_SHORT).show();
                btnStart.setEnabled(true);
            }
            else if (progressTortoise >= 100 &&progressHare < 100)
            {
                Toast.makeText(MainActivity.this, "The tortoise wins", Toast.LENGTH_SHORT).show();
                btnStart.setEnabled(true);
            }

            return false;
        }
    });

    private void runHare()
    {
        new Thread(() ->
        {
            Boolean[] sleepProb = {true, true, false};
            while (progressHare <= 100 && progressTortoise < 100)
            {
                try
                {
                    Thread.sleep(100);
                    if(sleepProb[(int) (Math.random() * 3)])
                        Thread.sleep(300);
                }
                catch (InterruptedException e)
                {
                    e.printStackTrace();
                }
                progressHare += 3;
                Message msg = new Message();
                msg.what = 1;
                handler.sendMessage(msg);
            }
        }).start();
    }

    private void runTortoise()
    {
        new Thread(() ->
        {
            while (progressTortoise <= 100 && progressHare < 100)
            {
                try
                {
                    Thread.sleep(100);
                }
                catch (InterruptedException e)
                {
                    e.printStackTrace();
                }
                progressTortoise += 1;

                Message msg = new Message();
                msg.what = 2;
                handler.sendMessage(msg);
            }
        }).start();
    }
}

File: Assignment3/KotlinThread1/app/src/main/java/com/example/kotlinthread1/MainActivity.kt
package com.example.kotlinthread1

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.os.Message
import android.widget.*
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

class MainActivity : AppCompatActivity()
{
    private var progressRabbit = 0
    private var progressTurtle = 0
    private lateinit var btnStart: Button
    private lateinit var sbHare: SeekBar
    private lateinit var sbTortoise: SeekBar
    override fun onCreate(savedInstanceState: Bundle?)
    {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        btnStart = findViewById(R.id.btnStart)
        sbHare = findViewById(R.id.sbHare)
        sbTortoise = findViewById(R.id.sbTortoise)
        btnStart.setOnClickListener {
            btnStart.isEnabled = false
            progressRabbit = 0
            progressTurtle = 0
            sbHare.progress = 0
            sbTortoise.progress = 0
            runRabbit()
            runTurtle()
        }
    }
    private fun showToast(msg: String) = Toast.makeText(this, msg, Toast.LENGTH_SHORT).show()

    private val handler = Handler(Looper.getMainLooper()) { msg ->
        if (msg.what == 1)
            sbHare.progress = progressRabbit
        if (progressRabbit >= 100 && progressTurtle < 100) {
            showToast("The hare wins!")
            btnStart.isEnabled = true
        }
        true
    }
    private fun runRabbit() {
        Thread {
            val sleepProbability = arrayOf(true, true, false)
            while (progressRabbit < 100 && progressTurtle < 100) {
                try
                {
                    Thread.sleep(100)
                    if (sleepProbability.random())
                        Thread.sleep(300)
                }
                catch (e: InterruptedException)
                {
                    e.printStackTrace()
                }
                progressRabbit += 3
                val msg = Message()
                msg.what = 1
                handler.sendMessage(msg)
            }
        }.start()
    }
    private fun runTurtle() {
        GlobalScope.launch(Dispatchers.Main)
        {
            while (progressTurtle < 100 && progressRabbit < 100)
            {
                delay(100)
                progressTurtle += 1
                sbTortoise.progress = progressTurtle
                if (progressTurtle >= 100 && progressRabbit < 100)
                {
                    showToast("The tortoise wins!")
                    btnStart.isEnabled = true
                }
            }
        }
    }
}